Allow disabling DropDownButton with optional hint for disabled state

escamoteur opened the pull request:
A quite common pattern in App that allows to filter Data, is selecting properties in multiple DropDownButtons where the available properties for the next DropDownbutton are only avaliable and dependending  on the value selected with the upper DropDownButton.
Currently it was not possible to disable a DropDownButton if you had no data ready for its items.. Therefore I make this PR.

I tried if it is possible to pass null or an empty List for `items` but both threw exceptions and did not disabling tapping on the button.

Now if you pass null for `items`it will disable the button and draw the down arrow in a lighter color.
 
Additionally you can provide an optional `hintDisabled' widget which will be displayed instead of `hint` if the button is disabled like here:

![image](https://user-images.githubusercontent.com/16635729/41601376-9481af7c-73d8-11e8-871b-23014287394f.png)

by updating `items` with a valid list it automatically get:

![image](https://user-images.githubusercontent.com/16635729/41603447-407e77fc-73dd-11e8-99c0-b847a1cfc3a6.png)


If no `hintDisabled` is provided `hint`will be displayed in disabled state too.

Hixie commented:
@escamoteur Looks like there was some issue where a formatter got a bit excited on this code; any chance you can undo the formatting changes to make the patch easier to review? Thanks.

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -464,6 +464,7 @@ class DropdownButton<T> extends StatefulWidget {
   /// Creates a dropdown button.
   ///
   /// The [items] must have distinct values and if [value] isn't null it must be among them.
+  /// If [items] is null the button is disabled and shows the optional [disabledHint] and the down arrow is drawn as disabled
```
nit: please wrap comments to about 80 characters and finish sentences with a period.

Hixie replied:
Consider something like:
```
  /// If [items] is null, the button will be disabled, the down arrow will be grayed out, and
  /// the [disabledHint] will be shown (if provided).
```

escamoteur replied:
done

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -472,14 +473,16 @@ class DropdownButton<T> extends StatefulWidget {
-      super(key: key);
+  })  : assert(items == null
+      || value == null
+      || items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
```
the formatting on this assert is a bit confusing. Consider something like:

```dart
  })  : assert(items == null || value == null ||
               items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
```

escamoteur replied:
done

Hixie commented:
I think it would be more idiomatic if the dropdown button was disabled when `onChanged` was null, rather than `items`.

Hixie commented:
This will need tests.

escamoteur commented:
Fixed the formatting as requested. I also added that if `onChanged == null` it disabled the control too. I think letting `items==null` make it disabled makes a lot of sense if you get this value updated by a `setState` quite often.

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -34,14 +34,14 @@ class _DropdownMenuPainter extends CustomPainter {
+      borderRadius: new BorderRadius.circular(2.0),
+      boxShadow: kElevationToShadow[elevation]
+  ).createBoxPainter(),
+        super(repaint: resize);
```
looks like you still have some indentation changes here

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -393,7 +393,7 @@ class DropdownMenuItem<T> extends StatelessWidget {
     @required this.child,
   }) : assert(child != null),
-       super(key: key);
+        super(key: key);
```
and some here and below

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -464,22 +464,25 @@ class DropdownButton<T> extends StatefulWidget {
-  }) : assert(items != null),
-       assert(value == null || items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
-      super(key: key);
+  })  : assert(items == null || value == null   || 
```
nit: extraneous spaces before the final `||`

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -492,6 +495,9 @@ class DropdownButton<T> extends StatefulWidget {
   /// Displayed if [value] is null.
   final Widget hint;
 
+  /// Displayed if [items] is null.
```
update documentation since it also displays if onChanged is null.

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -563,13 +569,17 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
-        return;
+    if (widget.items != null) {
+      assert(widget.value == null ||
+          widget.items
```
nit: indent this a bit more so it's vertically aligned (yes, the original code was bad for this too, oops)

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -613,27 +623,31 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
   }
 
+  // just so we don't have to repeat this test and make it easier to understand
+  bool get _enabled => widget.items != null && onChanged != null;
```
no need for the comment here

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -613,27 +623,31 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
     // item and the width of the hint.
-    final List<Widget> items = new List<Widget>.from(widget.items);
+    final List<Widget> items =
+    _enabled ? new List<Widget>.from(widget.items) : <Widget>[];
```
nit: indent by 4 if you wrap after a =

escamoteur replied:
done

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -644,17 +658,23 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
+                  : Colors.white70
+                  : Theme.of(context).brightness == Brightness.light
+                  ? Colors.grey.shade400
+                  : Colors.white10,
```
consider factoring this out into a separate method that uses "if" statements, it would be easier to read

escamoteur replied:
Done

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -673,7 +693,7 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
               height: 1.0,
               decoration: const BoxDecoration(
-                border: const Border(bottom: const BorderSide(color: const Color(0xFFBDBDBD), width: 0.0))
+                  border: const Border(bottom: const BorderSide(color: const Color(0xFFBDBDBD), width: 0.0))
```
nit: unnecessary whitespace change

escamoteur replied:
done

Hixie commented:
This will need tests.

Hixie commented:
(To write tests, look in the `//packages/flutter/test/material/` directory for the existing tests for this widget, and follow the same pattern.)

escamoteur commented:
I tried to add a test. unfortunately I could not run the test because on this branch other tests of this control are broken,

Hixie commented:
@escamoteur Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

escamoteur commented:
If you could take over that would be great, but I would be very disappointed if this would get lost after the work I done on it

Hixie commented:
Definitely won't get lost, we use the "needs love" label to look at closed PRs that we should be taking further. It just might take a long time, we don't have much time to give to these PRs yet. :-(

escamoteur commented:
@Hixie for now leave it open, I will try to get back to it in October

mklim commented:
I've adopted this into flutter/flutter#22539. Thanks for your contribution @escamoteur!

mklim commented:
Moved back from flutter/flutter#22539 to here, incorporated the early feedback from that PR.

Hixie commented:
LGTM modulo minor nits

Commit:
Squashed commit of the following:

commit 80bff41c4b19764f697458099cd46156794570cf
Author: Michael Klimushyn <mklim@google.com>
Date:   Mon Oct 15 10:45:03 2018 -0700

    Fix nits.

commit 2966ff99573025a82f67039ab3861a1190e3607b
Merge: ec73b47df 43001a3ab
Author: Michael Klimushyn <mklim@google.com>
Date:   Mon Oct 15 10:34:32 2018 -0700

    Merge branch 'master' into improved_dropdownbutton

commit ec73b47df08190c9b9ff68c4078fe2f2d0d5d04f
Author: Michael Klimushyn <mklim@google.com>
Date:   Tue Oct 2 10:44:41 2018 -0700

    Show disabled on onChanged or items null again.

    Also fixes tests.

commit 3a33eba9f7caca8d4c3e8773feac591759eba63e
Author: Michael Klimushyn <mklim@google.com>
Date:   Mon Oct 1 16:02:10 2018 -0700

    Dropdown shows disabledHint when items is empty

    See flutter/flutter#18770 for the original PR.

    Some changes in this patch:

    * Removed the behavior that showed this when `onChanged == null`. Most
    of dropdown's unit tests left `onChanged` as null and this broke them.
    Can put it back in and update them if that was just brittleness in the
    tests and we're not actively trying to preserve the previous behavior
    in that case.
    * Added the `items != null` assert back in. This functionality works
    just as well on an empty list as on null, and generally it's easier to
    reason about objects when they're guaranteed to be non-null.
    * Fix a bug where `disabledHint` wasn't being given enough padding,
    since `items` was empty.
    * Minor style changes.

commit cb069e2586c4c10c0bac953e67113355040126a7
Merge: 035ef2aa2 f19674853
Author: Michael Klimushyn <mklim@google.com>
Date:   Tue Oct 2 10:12:43 2018 -0700

    Merge branch 'master' into improved_dropdownbutton

commit 035ef2aa2798a82c526d4ef8d6dc41f8294f7653
Author: Michael Klimushyn <mklim@google.com>
Date:   Mon Oct 1 14:20:07 2018 -0700

    Revert whitespace changes.

commit 9f5c5aebdb911c57219f8959263cc0451ab81504
Author: escamoteur <github@burkharts.net>
Date:   Mon Jul 30 23:28:25 2018 +0200

    added test

commit 16d88ed420282c15d0885c3e2dfb9430fb122f7b
Author: escamoteur <github@burkharts.net>
Date:   Mon Jul 30 23:27:42 2018 +0200

    added test

commit baced964e784d40a80c42639355eb6c0140d4240
Author: escamoteur <github@burkharts.net>
Date:   Mon Jul 30 23:27:35 2018 +0200

    added the constraint that an empty items list disables the control too

commit 1ed6a6be9e766c7d042fe12630388c4236f73ba2
Author: escamoteur <github@burkharts.net>
Date:   Mon Jul 30 22:40:23 2018 +0200

    made all requested changes

commit f227bc221d4132528a89a01e325d0c0d6fb6e5f2
Author: escamoteur <github@burkharts.net>
Date:   Sun Jul 8 12:08:21 2018 +0200

    fixed formatting  and onChanged==null disabled the control too

commit 43957c4e16b6ae65e2d5dbf5ac9cf5a62efd8bb2
Author: escamoteur <github@burkharts.net>
Date:   Sun Jun 24 17:02:07 2018 +0200

    added disabled behaviour

mklim commented:
@escamoteur, do you mind if I force push a [squashed version of these changes](https://github.com/mklim/flutter/commit/2a7a7674538711104be9252e6873de8e8cc5b8f7) to this branch? The older commits and multiple merges are causing issues with how `flutter version` is picked up. Combining the commits and rebasing onto master fixes it, but will overwrite the existing history.

escamoteur commented:
Sure go on
